Common/Python/PythonInitializer.cs

```diff
@@ -57,31 +60,82 @@ public static void Initialize()
         {
-            if (paths == null)
+            // Filter out any paths that are already on our Python path
+            paths = paths?.Where(x => !pathCache.Contains(x.Replace('\\', '/'))).ToList();
+            if (paths == null || !paths.Any())
             {
                 return;
             }
 
+            // Add these paths to our pending additions
+            if (prepend)
+            {
+                _pendingStartOfPathAdditions.AddRange(paths);
+            }
+            else
+            {
+                _pendingEndOfPathAdditions.AddRange(paths);
+            }
+
             if (_isInitialized)
             {
                 using (Py.GIL())
                 {
-                    _pendingPathAdditions.AddRange(paths);
+                    dynamic sys = Py.Import("sys");
+                    var locals = new PyDict();
+                    locals.SetItem("sys", sys);
+
+                    // Insert any pending start of path additions
+                    if (!_pendingStartOfPathAdditions.IsNullOrEmpty())
+                    {
+                        PythonEngine.Exec(string.Join(";", _pendingStartOfPathAdditions.Select(s => $"sys.path.insert(0, '{s}')"))
+                            .Replace('\\', '/'), null, locals.Handle);
+
+                        _pendingStartOfPathAdditions.Clear();
+                    }
 
-                    // Generate the python code to add these to our path and execute
-                    var code = string.Join(";", _pendingPathAdditions.Select(s => $"sys.path.append('{s}')"))
-                        .Replace('\\', '/');
+                    // Append any end of path additions
+                    if (!_pendingEndOfPathAdditions.IsNullOrEmpty())
+                    {
+                        PythonEngine.Exec(string.Join(";", _pendingEndOfPathAdditions.Select(s => $"sys.path.append('{s}')"))
+                            .Replace('\\', '/'), null, locals.Handle);
 
-                    PythonEngine.Exec($"import sys;{code}");
-                    _pendingPathAdditions.Clear();
+                        _pendingEndOfPathAdditions.Clear();
+                    }
+
+                    // Update our path cache
+                    pathCache = PythonEngine.Eval("sys.path", null, locals.Handle).As<List<string>>();
+                    locals.Dispose();
                 }
             }
-            else
+        }
+
+        /// <summary>
+        /// "Activate" a virtual Python environment by prepending its library storage to PythonNets
+        /// path. This allows the libraries in this venv to be selected prior to our base install.
+        /// Requires PYTHONNET_PYDLL to be set to base install.
+        /// </summary>
+        /// <remarks>If a module is already loaded, Python will use its cached version first
+        /// these modules must be reloaded by reload() from importlib library</remarks>
```

Comment:
>       /// <remarks>If a module is already loaded, Python will use its cached version first
 >       /// these modules must be reloaded by reload() from importlib library</remarks>

Not sure we need to worry about supporting this use case? Should always be switch before starting? or maybe research use case 🤔

Reply:
Yeah little grey area here.

In the use case of launching lean, our implementation calls this before python is initialized so this problem does not apply. But when I was trying to use a unit test to prove the override, when the test assembly loaded it initialized PythonNet and imported pandas somewhere in our python project.

When it got to my unit test and I called this ActivatePythonVirtualEnvironment it worked on TF, since it hadn't been imported yet, but with Pandas it used the already loaded module.

Reply:
Just thought it may be nice to leave as a remark for anyone who messes around with this. 😁